### PR TITLE
Training throughput optimization: Fused Adam & not return logits

### PR DESCRIPTION
**Pull request by shehab-ashraf**
### Summary:
This PR introduces two key optimizations that improve training performance by 13% and fix an accuracy computation bug:

1. **Fused AdamW Optimizer**
2.  **Forward Pass** that computes loss and accuracy internally

### Results (L4 GPU):
| Metric               | Before            | After              | Improvement              |
|----------------------|-------------------|--------------------|--------------------------|
| Training Time        | 6m 15s            | 5m 28s             | 12.5% faster          |
| Throughput           | 14,729 tokens/s   | 16,654 tokens/s    |  13.1% faster          |
| Setup Time           | 195.58s           | 169s               |  13.6% faster          |
| Training Accuracy    | 0.020 (stuck!)     | 0.001 → 0.297      | Actually improves     |

>  **Bug Fixed**: Training accuracy was incorrectly computed (comparing predictions with unshifted labels), causing it to appear frozen at ~2%. It now correctly improves during training.

###  Changes Made:

1. Forward Pass:

#### Old Approach:
```
def forward(self, x):
    # ... model computation ...
    return logits

# In trainer:
logits = model(x)
shift_logits = logits[:, :-1, :].contiguous()
shift_labels = y[:, 1:].contiguous()
loss = F.cross_entropy(shift_logits.view(-1, vocab_size), shift_labels.view(-1))
# Accuracy computation (BUG: comparing with unshifted y!)
predictions = logits.argmax(dim=-1)
accuracy = (predictions == y).float().mean().item() 
```
#### New Approach:
```
def forward(self, x, labels=None):
    # ... model computation ...
    logits = self.lm_head(x)
    
    if labels is None:
        return logits 
    
    # Correctly shift for causal LM
    shift_logits = logits[:, :-1, :].contiguous()
    shift_labels = labels[:, 1:].contiguous()
    
    loss = nn.functional.cross_entropy(
        shift_logits.view(-1, self.config.vocab_size),
        shift_labels.view(-1)
    )
    
    # Accuracy on SHIFTED labels (correct!)
    predictions = shift_logits.argmax(dim=-1)
    accuracy = (predictions == shift_labels).float().mean()
    
    return {'loss': loss, 'accuracy': accuracy}
```

2. Fused AdamW Optimizer:
```
adamw_optimizer = torch.optim.AdamW(
    adamw_params,
    lr=config.adamw_lr,
    weight_decay=config.weight_decay,
    fused=torch.cuda.is_available()
)
```

### CODE REVIEW

**Comment by shehab-ashraf** [comment]
@vukrosic i see you only added Fused AdamW, what about forward pass with non return logits.
